client/network/src/request_responses.rs

```diff
@@ -0,0 +1,866 @@
+		for (p, _) in self.protocols.values_mut() {
+			NetworkBehaviour::inject_listener_closed(p, id, reason)
+		}
+	}
+
+	fn poll(
+		&mut self,
+		cx: &mut Context,
+		params: &mut impl PollParameters,
+	) -> Poll<
+		NetworkBehaviourAction<
+			<Self::ProtocolsHandler as ProtocolsHandler>::InEvent,
+			Self::OutEvent,
+		>,
+	> {
+		loop {
+			// Poll to see if any response is ready to be sent back.
+			// We need to check `is_empty` first, otherwise polling would return `None`.
+			if !self.pending_responses.is_empty() {
+				while let Poll::Ready(Some(outcome)) = self.pending_responses.poll_next_unpin(cx) {
+					match outcome {
+						RequestProcessingOutcome::PendingResponse {
+							protocol, inner_channel, response
+						} => {
+							if let Some((protocol, _)) = self.protocols.get_mut(&*protocol) {
+								protocol.send_response(inner_channel, response);
+							}
+						}
+						RequestProcessingOutcome::Busy { peer, protocol } => {
+							let out = Event::InboundRequest {
+								peer,
+								protocol,
+								outcome: Err(InboundError::Busy),
+							};
+							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+						}
+					}
+				}
+			}
+
+			// Poll request-responses protocols.
+			for (protocol, (behaviour, resp_builder)) in &mut self.protocols {
+				while let Poll::Ready(ev) = behaviour.poll(cx, params) {
+					let ev = match ev {
+						// Main events we are interested in.
+						NetworkBehaviourAction::GenerateEvent(ev) => ev,
+
+						// Other events generated by the underlying behaviour are transparently
+						// passed through.
+						NetworkBehaviourAction::DialAddress { address } => {
+							return Poll::Ready(NetworkBehaviourAction::DialAddress { address })
+						}
+						NetworkBehaviourAction::DialPeer { peer_id, condition } => {
+							return Poll::Ready(NetworkBehaviourAction::DialPeer {
+								peer_id,
+								condition,
+							})
+						}
+						NetworkBehaviourAction::NotifyHandler {
+							peer_id,
+							handler,
+							event,
+						} => {
+							return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+								peer_id,
+								handler,
+								event: ((*protocol).to_string(), event),
+							})
+						}
+						NetworkBehaviourAction::ReportObservedAddr { address } => {
+							return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+								address,
+							})
+						}
+					};
+
+					match ev {
+						// Received a request from a remote.
+						RequestResponseEvent::Message {
+							peer,
+							message: RequestResponseMessage::Request { request, channel },
+						} => {
+							let (tx, rx) = oneshot::channel();
+
+							// Submit the request to the "response builder" passed by the user at
+							// initialization.
+							if let Some(resp_builder) = resp_builder {
+								// If the response builder is too busy, silently drop `tx`.
+								// This will be reported as a `Busy` error.
```

Comment:
The requirement to size the "response builder" channel properly and the dropping of requests when it has reached capacity look quite brittle to me. I would expect back-pressure is exercised, i.e. unless the response builder is ready to process another inbound request, the `RequestResponseBehaviour` should not even accept new inbound requests, such that the senders can not produce ever more. I do realise that the events from `RequestResponse` need to be processed, but maybe the back-pressure handling should be rethought, e.g. instead of immediately passing on the whole inbound request `RequestResponse` may queue a single request, generate an event that a request is ready to be consumed and only after `RequestResponseBehaviour` actually retrieved it will `RequestResponse` read another inbound request.

Reply:
It's not really a question of API here, but of what to do on a more fundamental level.

If we are connected to 200 other nodes, and they all send us a request at the same time, and it takes on average 100ms to build a response to this request, then the 200th node would receive a response only after 20 seconds.

This PR assumes that 20 seconds is an unacceptably long time to send back a response, and prefers to immediately make further requests fail rather than proceeding to answer them.

Reply:
I agree this is a fundamental decision, but my conclusion is the opposite to yours. In particular I think it is up to the requesting endpoint to decide how long it is willing to wait for an answer.

Reply:
On a general level, I think it would be nice if a configurable concurrent inbound substream limit could be employed for back-pressure already in the muxers, but as @twittner pointed out, that is probably not how the muxers currently behave. Of course, such a limit would also apply to all substreams (whether used for notifications, request-response, etc.).

I think all such measures - not accepting new substreams (or delaying accepting them), sending early error responses, "dropping" requests (i.e. just closing the substream, a form of inbound substream limit at a higher layer) - are reasonable and can be complementary for managing inbound traffic. I just also think that considerations about average response times are not really that important for sizing queues on the receiving end, as the numbers involved are subject to a lot of variation. Rather, the receiver merely decides how much resources (primarily in terms of memory) it is willing to dedicate to pending requests for a certain protocol. With the `max_request_size` one can calculate the upper limits reasonably well. How long one is willing to wait for a response, on the other hand, is primarily up to the sender of a request and different senders can have different tolerances in that regard - one shouldn't assume that all senders have the same request timeout, since we're not even assuming all peers to be substrate nodes.

Reply:
I generally agree with @twittner's suggestion to modify the way the `RequestResponse` behaves.

However, muxers in general indeed don't permit applying back-pressure on incoming substreams. As far as I know, it's not something that we can change in rust-libp2p, but rather something that multiplexing protocols in general don't permit.

I think that there are two possible courses of actions:

- Change the protocol to no longer open a substream per request, but rather append multiple requests one behind the other in the same substream.
- Enforce a limit, in the protocol, to the number of ongoing requests. The sending side would have to throttle itself in order to not reach this limit.

Reply:
The throttling can be made explicit in the protocol. If we augment the request-response protocol and send metadata along with the actual payload data (something we should maybe anyway do for future extensibility) we could make use of tokens which are required for sending. More specifically, the receiver issues tokens to the sender which are valid for a given number of requests the sender is allowed to send before having to wait for the next token. The initial request is free because the sender does not have a token yet and we probably do not want to require a roundtrip to get one, otherwise we could require that the sender first has to get a token. `RequestResponse` queues the requests and issues a new token to the sender when the queue is empty:

```
   Sender                     Receiver                           Consumer
|         |               |             |                       |
|Request  |----(Send)---->|[]           |                       |
|         |               |             |                       |
|<wait>   |               |[Request]    |<---(Get Request)----- |
|         |               |             |<-(Accept N requests)--|
|         |               |             |                       |
|         |<-(Token1(N))--|[]           |                       |
|         |               |             |                       |
|Request1,|               |             |                       |
|Token1   |----(Send)---->|[Request1]   |                       |
|         |               |             |                       |
|...      |               |             |                       |
|         |               |             |                       |
|RequestN,|               |             |                       |
|Token1   |----(Send)---->|[Request1,   |                       |
|         |               | ... RequestN|                       |
|         |               |]            |                       |
|         |               |             |                       |
|<wait>   |               |[Request1,   |<---(Get Request)------|
|         |               | ... RequestN|                       |
|         |               |]            |                       |
|         |               |             |                       |
|         |               |...          |                       |
|         |               |             |                       |
|         |               |[RequestN]   |<---(Get Request)------|
|         |               |             |<-(Accept M requests)--|
|         |<-(Token2(M))--|[]           |                       |
...
```
Sending requests with invalid tokens or tokens whose associated quota is used up is a protocol violation. The quota applies per peer and protocol and is independent on the number of substreams or connections.

Reply:
I still don't know what to do, practically speaking. As a reminder, we're trying to ship parachains as soon as possible, and I would like to avoid this protocol design discussion taking two months of back-and-forth.

Reply:
I started prototyping the required changes here: https://github.com/tomaka/polkadot/compare/request-responses...twittner:budgets It contains some TODOs, mostly to reset send/receive budgets on timeouts etc. Let me know if this makes sense to you.

Reply:
I'm quite reluctant to implement something complex, as it means:

- We're potentially going to spend a lot of time implementing it and fixing bugs. Same for any other implementation team.
- We wouldn't be able to base the existing block-request/finality-request/light-client-requests on top of this new request-response system unless we do a transition period, which is again quite annoying.
- I'm not convinced that there doesn't exist a more simple solution that achieves the same. In particular, is there anything wrong with instead limiting the number of simultaneous requests per peer for any given protocol?

Reply:
> In particular, is there anything wrong with instead limiting the number of simultaneous requests per peer for any given protocol?

As discussed in private, a protocol level constant should be fine and does not require the proper sizing of mpsc channels. It is awkward to roll out changes to this constant but as I understood it this is not expected to happen.